integtest/spec/single_book_spec.rb

```diff
@@ -0,0 +1,209 @@
+      --doc #{from}
+      --out #{to}
+    ]
+  end
+
+  ##
+  # Context for converting asciidoc files into html. To use this include it in
+  # a context and then define a `make_input` method that initializes the source
+  # files and returns the path of the "index" asciidoc file, preferably
+  # using `write_source`.
+  shared_context 'convert' do
+    include_context 'tmp dirs'
+    before(:context) do
+      from = make_input
+      init_repo(File.expand_path('..', from))
+      # Use popen here instead of capture to keep stdin open to appease the
+      # docker-image-always-removed paranoia in build_docs.pl
+      _stdin, out, wait_thr = Open3.popen2e(
+        '/docs_build/build_docs.pl', '--in_standard_docker',
+        *convert_args(from, @dest)
+      )
+      status = wait_thr.value
+      raise_status cmd, out, status unless status.success?
+
+      out
+    end
+    it 'creates the template hash' do
+      expect(dest_file('template.md5')).to file_exist
+    end
+    it 'creates the css' do
+      expect(dest_file('styles.css')).to file_exist
+    end
+    it 'creates the js' do
+      expect(dest_file('docs.js')).to file_exist
+    end
+  end
+
+  ##
+  # Include in a context named after a page to add an assertion that the page
+  # was created and to create some variables extracted from the page's contents
+  #
+  # body - the non-template body of the page
+  # title - the title of the page
+  shared_context 'page body' do
+    let(:file) do |example|
+      dest_file(example.example_group.description)
+    end
+    it 'is created' do
+      expect(file).to file_exist
+    end
+    let(:body) do
+      return nil unless File.exist? file
+
+      File.open(dest_file(file), 'r:UTF-8') do |f|
+        f.read
+         .sub(/.+<!-- start body -->/m, '')
+         .sub(/<!-- end body -->.+/m, '')
+      end
+    end
+    let(:title) do
+      return nil unless body
+
+      m = body.match(%r{<h1 class="title"><a id=".+"></a>([^<]+)(<a.+?)?</h1>})
+      raise "Can't find title in #{body}" unless m
+
+      m[1]
+    end
+  end
+
+  HEADER = <<~ASCIIDOC
+    = Title
+
+    [[chapter]]
+    == Chapter
+  ASCIIDOC
+
+  context 'for a minimal book' do
+    shared_context 'expected' do
+      def make_input
+        write_source file_name, <<~ASCIIDOC
+          #{HEADER}
+          This is a minimal viable asciidoc file for use with build_docs. The
+          actual contents of this paragraph aren't important but having a
+          paragraph here is required.
+        ASCIIDOC
+      end
+      include_context 'convert'
+
+      context 'index.html' do
+        include_context 'page body'
+        it 'has the right title' do
+          expect(title).to eq('Title')
+        end
+      end
+      context 'chapter.html' do
+        include_context 'page body'
+        it 'has the right tile' do
+          expect(title).to eq('Chapter')
+        end
+      end
+    end
+
+    context 'when the file ends in .asciidoc' do
+      def file_name
```

Comment:
It seems strange to me to use a method here instead of a `let` block. Is there a reason for this?

Reply:
`let` block happen too late! The `begin` block can't see the input.

I wonder if I could do this another way, maybe replacing the `convert` shared context with something like:
```
def convert_context
    before(:context) do
      from = yield
      init_repo(File.expand_path('..', from))
      # Use popen here instead of capture to keep stdin open to appease the
      # docker-image-always-removed paranoia in build_docs.pl
      _stdin, out, wait_thr = Open3.popen2e(
        '/docs_build/build_docs.pl', '--in_standard_docker',
        *convert_args(from, @dest)
      )
      status = wait_thr.value
      raise_status cmd, out, status unless status.success?

      out
    end
    ...
end
```